include/astshim/FrameDict.h

```diff
@@ -0,0 +1,310 @@
+                   [](unsigned char c) -> unsigned char { return std::toupper(c); });
+        auto it = _domainIndexDict.find(domainUppercase);
+        if (it == _domainIndexDict.end()) {
+            throw std::range_error("No frame found with domain " + domain);
+        }
+        return it->second;
+    }
+
+    /**
+    Return True if a frame in this FrameDict has the specified domain
+    */
+    bool hasDomain(std::string const &domain) const { return bool(_domainIndexDict.count(domain) > 0); }
+
+    using FrameSet::mirrorVariants;
+
+    /**
+    Call mirrorVariants with a frame specified by domain
+
+    @throw std::range_error if no frame found with the specified domain
+    */
+    void mirrorVariants(std::string const &domain) { FrameSet::mirrorVariants(getIndex(domain)); }
+
+    using FrameSet::remapFrame;
+
+    /**
+    Remap a frame specified by domain
+
+    @throw std::range_error if no frame found with the specified domain
+    */
+    void remapFrame(std::string const &domain, Mapping &map) { FrameSet::remapFrame(getIndex(domain), map); }
+
+    using FrameSet::removeFrame;
+
+    /**
+    Remove a frame specified by domain
+
+    @throw std::range_error if no frame found with the specified domain
+    */
+    void removeFrame(std::string const &domain) {
+        FrameSet::removeFrame(getIndex(domain));
+        _rebuildDict();
+    }
+
+    using FrameSet::setBase;
+
+    /**
+    Set the base frame to a frame specified by domain
+
+    @throw std::range_error if no frame found with the specified domain
+    */
+    void setBase(std::string const &domain) { FrameSet::setBase(getIndex(domain)); }
+
+    using FrameSet::setCurrent;
+
+    /**
+    Set the current frame to a frame specified by domain
+
+    @throw std::range_error if no frame found with the specified domain
+    */
+    void setCurrent(std::string const &domain) { FrameSet::setCurrent(getIndex(domain)); }
+
+    /// Set the domain of the current frame (and update the internal dict)
+    void setDomain(std::string const &domain) override {
+        Frame::setDomain(domain);
+        _rebuildDict();
+    }
+
+protected:
+    virtual std::shared_ptr<Object> copyPolymorphic() const override {
+        return copyImpl<FrameDict, AstFrameSet>();
+    }
+
+    /// Construct a FrameDict from a raw AST pointer
+    explicit FrameDict(AstFrameSet *rawptr) : FrameSet(rawptr) {
+        if (!astIsAFrameSet(getRawPtr())) {
+            std::ostringstream os;
+            os << "this is a " << getClassName() << ", which is not a FrameSet";
+            throw std::invalid_argument(os.str());
+        }
+        _rebuildDict();
+    }
+
+private:
+    // Rebuild the internal domain:index dictionary
+    void _rebuildDict() {
+        _domainIndexDict.clear();
+        for (int index = 1, end = getNFrame(); index <= end; ++index) {
+            auto const frame = FrameSet::getFrame(index, false);
+            _addFrameToDict(*frame, index);
+        }
```

Comment:
Yes, but I think asserts will do fine here. Throwing an exception is fine for a constructor (since the state is irrelevant in that case) and asserts are good enough the rest of the time.